bundle/jsky.app.ot/src/main/scala/jsky/app/ot/tpe/feat/TpeGhostIFUFeature.scala

```diff
@@ -0,0 +1,472 @@
+import java.beans.{PropertyChangeEvent, PropertyChangeListener}
+import java.util.Collections
+
+import edu.gemini.pot.sp.SPComponentType
+import edu.gemini.shared.util.immutable.{None => JNone, Option => JOption, Some => JSome}
+import edu.gemini.shared.util.immutable.ScalaConverters._
+import edu.gemini.spModel.core.Angle
+import edu.gemini.spModel.gemini.ghost.{Ghost, GhostAsterism, GhostScienceAreaGeometry}
+import edu.gemini.spModel.obs.context.ObsContext
+import edu.gemini.spModel.obscomp.SPInstObsComp
+import edu.gemini.spModel.target.{SPSkyObject, WatchablePos}
+import edu.gemini.spModel.target.env.{Asterism, AsterismType, TargetEnvironment, TargetEnvironmentDiff}
+import javax.swing.{Icon, JLabel, JPanel, SwingConstants}
+import jsky.app.ot.tpe._
+import jsky.app.ot.util.{BasicPropertyList, OtColor, PropertyWatcher}
+
+import scala.swing.{Color, Graphics2D}
+
+import scalaz._
+import Scalaz._
+
+import edu.gemini.spModel.gemini.ghost.GhostAsterism.GhostTarget
+
+/**
+ * Draws the GHOST IFU patrol fiels and IFUS.
+ */
+final class TpeGhostIfuFeature extends TpeImageFeature("GHOST", "Show the patrol fields of the GHOST IFUs") with PropertyWatcher with TpeDragSensitive {
+  // Transformations for the GHOST IFU patrol fields.
+  private var transform1: Option[AffineTransform] = None
+  private var transform2: Option[AffineTransform] = None
+  private var ifuTransform: Option[AffineTransform] = None
+
+  // Determine if there are no TPE messages.
+  private var isEmpty: Boolean = false
+
+  /**
+   * Determine which IFUs are active, e.g. when targets / sky positions are being dragged.
+   * This should be set by the handle methods. This does not determine if both the IFU patrol fields are drawn
+   * in and of itself alone, as we only draw the IFU2 elements if it is in use.
+   */
+  private var displayMode: TpeGhostIfuFeature.IFUDisplayMode = TpeGhostIfuFeature.IFUDisplayModeBoth
+
+  /**
+   * If _flipRA is -1, flip the RA axis of the area.
+   */
+  private def flipArea(a: Area): Area =
+    if (_flipRA == -1) a.createTransformedArea(AffineTransform.getScaleInstance(_flipRA, 1.0))
+    else a
+
+  /**
+   * Draw the feature: two overlapping symmetric arcs.
+   */
+  override def draw(g: Graphics, tii: TpeImageInfo): Unit = {
+    if (!isEnabled(_iw.getContext))
+      return
+
+    for {
+      trans1 <- transform1
+      trans2 <- transform2
+      ctx <- _iw.getObsContext.asScalaOpt
+    } {
+      // Store the old color and paint.
+      val g2d: Graphics2D = g.asInstanceOf[Graphics2D]
+      val originalColor = g2d.getColor
+      val originalPaint = g2d.getPaint
+
+      val env: TargetEnvironment = ctx.getTargets
+      if (env == null)
+        return
+
+      if (drawPatrolFields) {
+        val ifu1PatrolField: Area = new Area(flipArea(TpeGhostIfuFeature.IfuArc)).createTransformedArea(trans1)
+        val ifu2PatrolField: Area = new Area(flipArea(TpeGhostIfuFeature.IfuArc)).createTransformedArea(trans2)
```

Comment:
Clearly, this is wrong. Just to clarify:
1. If there is an offset, the patrol field should be drawn relative to the offset.
2. If there are multiple offsets, the patrol fields should be drawn relative to the intersections of the offsets.
That's correct, right? I'm sorry I got the offsets wrong again. I know this is not the first time. Thanks for catching this and I apologize for you having to find it.

Reply:
> If there is an offset, the patrol field should be drawn relative to the offset.

That's right, at least when that offset is selected in the Offset iterator in the OT.  It can be kind of flakey.

> If there are multiple offsets, the patrol fields should be drawn relative to the intersections of the offsets.

Right, we take the intersection of the patrol field at each offset and get a shape (which may be empty if the offsets are too far apart).